File: src/main/java/com/bvanseg/gigeresque/client/entity/render/feature/EggmorphFeatureRenderer.java
package com.bvanseg.gigeresque.client.entity.render.feature;

import java.util.HashMap;

import com.bvanseg.gigeresque.Constants;
import com.bvanseg.gigeresque.client.entity.texture.EggmorphLayerTexture;
import com.bvanseg.gigeresque.interfacing.Eggmorphable;

import net.minecraft.client.MinecraftClient;
import net.minecraft.client.render.OverlayTexture;
import net.minecraft.client.render.VertexConsumerProvider;
import net.minecraft.client.render.entity.feature.FeatureRenderer;
import net.minecraft.client.render.entity.feature.FeatureRendererContext;
import net.minecraft.client.render.entity.model.EntityModel;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.entity.Entity;
import net.minecraft.util.Identifier;

public class EggmorphFeatureRenderer<T extends Entity, M extends EntityModel<T>> extends FeatureRenderer<T, M> {
	public EggmorphFeatureRenderer(FeatureRendererContext<T, M> context) {
		super(context);
	}

	private static final HashMap<Identifier, EggmorphLayerTexture> textureCache = new HashMap<Identifier, EggmorphLayerTexture>();

	@SuppressWarnings("resource")
	public static <T extends Entity> void renderEggmorphedModel(EntityModel<T> renderedModel, Identifier texture,
			MatrixStack matrices, VertexConsumerProvider vertexConsumers, int light, T entity, float limbAngle,
			float limbDistance, float tickDelta, float animationProgress, float headYaw, float headPitch) {
		matrices.push();
		renderedModel.animateModel(entity, limbAngle, limbDistance, tickDelta);
		var vertexConsumer = vertexConsumers.getBuffer(getEggmorphLayerTexture(texture).renderLayer);
		var progress = 1 - ((Eggmorphable) entity).getTicksUntilEggmorphed() / ((float) Constants.EGGMORPH_DURATION);
		renderedModel.setAngles(entity, limbAngle, limbDistance, animationProgress, headYaw, headPitch);
		renderedModel.render(matrices, vertexConsumer, light, OverlayTexture.DEFAULT_UV, 1F, 1F, 1F, progress);
		matrices.pop();
	}

	public static EggmorphLayerTexture getEggmorphLayerTexture(Identifier texture) {
		return textureCache.computeIfAbsent(texture,
				identifier -> new EggmorphLayerTexture(MinecraftClient.getInstance().getTextureManager(),
						MinecraftClient.getInstance().getResourceManager(), texture));
	}

	@Override
	public void render(MatrixStack matrices, VertexConsumerProvider vertexConsumers, int light, T entity,
			float limbAngle, float limbDistance, float tickDelta, float animationProgress, float headYaw,
			float headPitch) {
		if (entity instanceof Eggmorphable && ((Eggmorphable) entity).isEggmorphing()) {
			renderEggmorphedModel(getContextModel(), getTexture(entity), matrices, vertexConsumers, light, entity,
					limbAngle, limbDistance, tickDelta, animationProgress, headYaw, headPitch);
		}
	}
}

File: src/main/kotlin/com/bvanseg/gigeresque/client/entity/render/feature/EggmorphFeatureRenderer.kt
package com.bvanseg.gigeresque.client.entity.render.feature

import com.bvanseg.gigeresque.Constants
import com.bvanseg.gigeresque.client.entity.texture.EggmorphLayerTexture
import com.bvanseg.gigeresque.interfacing.Eggmorphable
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.client.MinecraftClient
import net.minecraft.client.render.OverlayTexture
import net.minecraft.client.render.VertexConsumerProvider
import net.minecraft.client.render.entity.feature.FeatureRenderer
import net.minecraft.client.render.entity.feature.FeatureRendererContext
import net.minecraft.client.render.entity.model.EntityModel
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.entity.Entity
import net.minecraft.util.Identifier

/**
 * @author Aelpecyem
 */
@Environment(EnvType.CLIENT)
class EggmorphFeatureRenderer<T : Entity, M : EntityModel<T>>(context: FeatureRendererContext<T, M>) :
    FeatureRenderer<T, M>(context) {

    companion object {
        private val textureCache = HashMap<Identifier, EggmorphLayerTexture>()

        @JvmStatic
        fun <T : Entity> renderEggmorphedModel(
            renderedModel: EntityModel<T>,
            texture: Identifier,
            matrices: MatrixStack,
            vertexConsumers: VertexConsumerProvider,
            light: Int,
            entity: T,
            limbAngle: Float,
            limbDistance: Float,
            tickDelta: Float,
            animationProgress: Float,
            headYaw: Float,
            headPitch: Float
        ) {
            matrices.push()
            renderedModel.animateModel(entity, limbAngle, limbDistance, tickDelta)
            val vertexConsumer = vertexConsumers.getBuffer(
                getEggmorphLayerTexture(
                    texture
                ).renderLayer
            )
            val progress = 1 - (entity as Eggmorphable).ticksUntilEggmorphed / (Constants.EGGMORPH_DURATION.toFloat())
            renderedModel.setAngles(entity, limbAngle, limbDistance, animationProgress, headYaw, headPitch)
            renderedModel.render(matrices, vertexConsumer, light, OverlayTexture.DEFAULT_UV, 1F, 1F, 1F, progress)
            matrices.pop()
        }

        @JvmStatic
        fun getEggmorphLayerTexture(texture: Identifier): EggmorphLayerTexture {
            return textureCache.computeIfAbsent(
                texture
            ) {
                EggmorphLayerTexture(
                    MinecraftClient.getInstance().textureManager,
                    MinecraftClient.getInstance().resourceManager,
                    texture
                )
            }
        }
    }

    override fun render(
        matrices: MatrixStack,
        vertexConsumers: VertexConsumerProvider,
        light: Int,
        entity: T,
        limbAngle: Float,
        limbDistance: Float,
        tickDelta: Float,
        animationProgress: Float,
        headYaw: Float,
        headPitch: Float
    ) {
        if (entity is Eggmorphable && entity.isEggmorphing) {
            renderEggmorphedModel(
                contextModel,
                getTexture(entity),
                matrices,
                vertexConsumers,
                light,
                entity,
                limbAngle,
                limbDistance,
                tickDelta,
                animationProgress,
                headYaw,
                headPitch
            )
        }
    }
}